### PR TITLE
Upgrade to declarative setuptools

### PR DESCRIPTION
**Pull request by graingert**

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -1,2 +1,39 @@
-[wheel]
-universal=1
+[metadata]
+name = incremental
+version = attr: incremental._setuptools_version
```
Comment (altendky) [inline setup.cfg:3]: This seems like a cleaner example to give if it works.  If taking this, remove the `def` added down below.

```suggestion
version = attr: incremental.__version__.public
```

File: src/incremental/__init__.py [hunk 2]
```diff
@@ -391,4 +391,8 @@ def _get_version(dist, keyword, value):
 from ._version import __version__  # noqa
 
 
+def _setuptools_version():
+    return __version__.public()
+
+
```
Comment (altendky) [inline src/incremental/__init__.py:397]: Interesting, the previously used `.base()` is one of three aliases for `.public()`...

https://github.com/twisted/incremental/blob/8fceca84f4d677ac497174025f11e25b6abbf110/src/incremental/__init__.py#L245-L247

If taking the suggestion above then remove this.

```suggestion
```

**Review by altendky — CHANGES_REQUESTED** [review]
Probably at least need to have coverage of Python 2 to verify it all works appropriately with whatever pip Python 2 gets to use.  I won't be picky about where that runs for the sake of this.

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,3 @@
+[build-system]
+requires = [ "setuptools >= 35.0.2", "wheel >= 0.29.0"]
```
Comment (graingert) [inline pyproject.toml:2]: I think I copied them from somewhere else, lets get the best versions that py2.7 supports:

```suggestion
requires = [ "setuptools >= 44.1.1", "wheel >= 0.36.2"]
```

File: setup.cfg (hunk 1 above)
Comment (graingert) [reply setup.cfg:3]: ModuleNotFoundError: No module named 'incremental.__version__'

File: src/incremental/__init__.py (hunk 2 above)
Comment (graingert) [reply src/incremental/__init__.py:397]: yeah it doesn't work, sadly

File: setup.cfg (hunk 1 above)
Comment (altendky) [reply setup.cfg:3]: Welp, definitely a sign of a thing to fix as has been discussed on IRC.

**Review by altendky — COMMENTED** [review]
Good to go once testing on both Python 2 and 3 is in place.

File: setup.cfg [hunk 4]
```diff
@@ -1,2 +1,39 @@
...
+url = https://github.com/twisted/incremental
+classifiers =
+    Intended Audience :: Developers
+    License :: OSI Approved :: MIT License
+    Programming Language :: Python :: 2
+    Programming Language :: Python :: 2.7
+    Programming Language :: Python :: 3
+    Programming Language :: Python :: 3.4
+    Programming Language :: Python :: 3.5
+    Programming Language :: Python :: 3.6
```
Comment (altendky) [inline setup.cfg:15]: Based on https://github.com/twisted/incremental/pull/70 we can add a few.

```suggestion
    Programming Language :: Python :: 3.6
    Programming Language :: Python :: 3.7
    Programming Language :: Python :: 3.8
    Programming Language :: Python :: 3.9
```

**Pull request merged**